File: ingraph-compiler/ingraph-compiler-relalg-model-util/src/main/java/ingraph/relalg/collectors/CollectionHelper.java
package ingraph.relalg.collectors;

import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;
import com.google.common.collect.Sets;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.function.Consumer;
import org.eclipse.xtext.xbase.lib.Functions.Function1;
import org.eclipse.xtext.xbase.lib.IterableExtensions;
import org.eclipse.xtext.xbase.lib.ListExtensions;
import relalg.RelalgModelElement;

@SuppressWarnings("all")
public class CollectionHelper {
  public <T extends Object> List<T> union(final Iterable<? extends T>... lists) {
    Iterable<T> _concat = Iterables.<T>concat(lists);
    return Lists.<T>newArrayList(_concat);
  }
  
  public <T extends RelalgModelElement> List<T> uniqueUnion(final Iterable<? extends T>... lists) {
    ArrayList<T> _xblockexpression = null;
    {
      final ArrayList<T> result = new ArrayList<T>();
      List<T> _union = this.<T>union(lists);
      final Consumer<T> _function = new Consumer<T>() {
        public void accept(final T it) {
          final Function1<T, String> _function = new Function1<T, String>() {
            public String apply(final T it) {
              return it.fullName();
            }
          };
          List<String> _map = ListExtensions.<T, String>map(result, _function);
          String _fullName = it.fullName();
          boolean _contains = _map.contains(_fullName);
          boolean _not = (!_contains);
          if (_not) {
            result.add(it);
          }
        }
      };
      _union.forEach(_function);
      _xblockexpression = result;
    }
    return _xblockexpression;
  }
  
  public <T extends Object> List<T> minus(final Iterable<? extends T> iterable1, final Iterable<? extends T> iterable2) {
    List<T> _xblockexpression = null;
    {
      final LinkedHashSet<T> set1 = Sets.<T>newLinkedHashSet(iterable1);
      final LinkedHashSet<T> set2 = Sets.<T>newLinkedHashSet(iterable2);
      Sets.SetView<T> _difference = Sets.<T>difference(set1, set2);
      _xblockexpression = IterableExtensions.<T>toList(_difference);
    }
    return _xblockexpression;
  }
}

File: ingraph-compiler/ingraph-compiler-relalg-model-util/src/main/scala/ingraph/relalg/collectors/CollectionHelper.scala
package ingraph.relalg.collectors

import java.{lang, util}

import com.google.common.collect.{ImmutableList, Iterables, Lists, Sets}
import relalg.RelalgModelElement

import scala.collection.JavaConverters._
import scala.collection.mutable.ListBuffer

class CollectionHelper {

  def union_n[T](lists: java.lang.Iterable[_ <: T]*): java.util.List[T] = {
    new util.ArrayList(lists.map(x => x.asScala).flatten.toList.asJava)
  }

  def union[T](l1: java.lang.Iterable[_ <: T], l2: java.lang.Iterable[_ <: T]): java.util.List[T] = {
    union_n(l1, l2)
  }

  def union[T](l1: java.lang.Iterable[_ <: T], l2: java.lang.Iterable[_ <: T], l3: java.lang.Iterable[_ <: T]): java.util.List[T] = {
    union_n(l1, l2, l3)
  }

  def uniqueUnion_n[T <: RelalgModelElement](lists: java.lang.Iterable[_ <: T]*): java.util.List[T] = {
    val result = new ListBuffer[T]

    union_n(lists: _*).asScala.foreach(x =>
      if (!result.map(_.fullName).contains(x.fullName)) {
        result += x
      }
    )

    new util.ArrayList(result.asJava)
  }

  def uniqueUnion[T <: RelalgModelElement](l1: java.lang.Iterable[_ <: T], l2: java.lang.Iterable[_ <: T]): java.util.List[T] = {
    uniqueUnion_n(l1, l2)
  }

  def uniqueUnion[T <: RelalgModelElement](l1: java.lang.Iterable[_ <: T], l2: java.lang.Iterable[_ <: T], l3: java.lang.Iterable[_ <: T]): java.util.List[T] = {
    uniqueUnion_n(l1, l2, l3)
  }

  def minus[T](i1: lang.Iterable[_ <: T], i2: lang.Iterable[_ <: T]): java.util.List[T] = {
    val x : java.util.Set[T] = Sets.difference(
      Sets.newLinkedHashSet(i1),
      Sets.newLinkedHashSet(i2)
    )

    new util.ArrayList(x)
  }

}
